geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationEventQueueManager.java

```diff
@@ -99,6 +97,132 @@ void add(final InternalCacheEvent event,
     }
   }
 
+  void drain(final ClientRegistrationEventQueue clientRegistrationEventQueue,
+      final CacheClientNotifier cacheClientNotifier) {
+    try {
+      ClientProxyMembershipID clientProxyMembershipID =
+          clientRegistrationEventQueue.getClientProxyMembershipID();
+
+      if (logger.isDebugEnabled()) {
+        logger.debug("Draining events from registration queue for client proxy "
+            + clientProxyMembershipID
+            + " without synchronization");
+      }
+
+      CacheClientProxy cacheClientProxy = cacheClientNotifier
+          .getClientProxy(clientProxyMembershipID);
+
```

Comment:
ryan, do you feel that a null check here will avoid couple of calls and locks ? I mean if cache client proxy is null we just decrement the put in progress counter and return from here.

Reply:
So we still need to do the locking so that no new events are added while we are processing and removing the queue.  We need to decrement the put in progress counter on every event in the queue, and also need to ensure no new events are being added while we do that (otherwise we would remove the queue from the manager while it may still have events in it, thereby resulting in events leaking without the counter being decremented).  If you look at the drain logic, the only thing it does is decrement the putinprogress counter for each event if the proxy is null, so we aren't really doing any unnecessary work.  But we do need to do that for every event in the queue under synchronization.

Reply:
I think maybe the confusion here is that the put in progress counter is incremented _per event received_ during registration, not just once for the whole registration.

Reply:
I added a new test to ensure the put in progress counter is properly incremented and decremented for each event.  This conversation made me realize that we need a test for that case.